### ISSUE TITLE
[Query] Failure of multi-part message in JeroMQ

### ISSUE THREAD

**Issue by jay11ca39** [zeromq/jeromq#519]
Hello All,
This issue is created to know about internals of multi-part message.
**I am creating a publisher socket and i am sending multi-part message using sendMore API.**

**Question:**

1. Suppose, I have publish API which will publish the given message on pub socket.

2. Application want to send message of 4 frames. So, my publish API has to call sendMore() API 3 times and then at last send() API.
Suppose in second call to sendMore() API fails. And then, application again call publish API with new message. then again publish API will again call sendMore() API 3 times and then at last send() API for the new message. this time it succeed. 

Will internally JeroMQ will have 1 old frame + 4 new frames and it will send all the five frames?
Or there is some other logic which will prevent this kind of situation?

**Comment by daveyarwood** [comment]
I took a brief look at the code and I have some observations about what it looks like the code is doing:

- `sendMore` will return `true` if successful.
- If sending fails with an `errno` of `ZError.EAGAIN`, `false` is returned.
- If sending fails with any other error, the logic gets rather [hard to follow](https://github.com/zeromq/jeromq/blob/225a521d12c40f0fe6a01e0cdd001c1f8366b2f5/src/main/java/zmq/SocketBase.java#L722-L759) and depends on a variety of factors including the sendTimeout option and whether the `ZMQ.ZMQ_DONTWAIT` bit flag is set.

It does appear that JeroMQ will retry up until the timeout if the error is not EAGAIN. It isn't really clear to me if this will result in the messages being sent out of order, or what exactly a "command" is -- it appears that before re-sending, JeroMQ will "wait for the next command, process it and try to send the message again." up until the timeout, if any.

**Comment by daveyarwood** [comment]
> as per my understanding it will queue till final send api is called

I don't think that's true. I think it sends the messages as you send them, and if you use `sendMore`, then the receiver can observe that there are more frames to receive (via `hasReceiveMore`) and keep receiving them.

**Comment by fredoboulo** [comment]
Hello,

Could you explain how or in which of your circumstances sendMore() call would fail on a PUB socket?

My understanding is that the pipes accumulate messages with MORE flag, until the final one is sent, which flushes the pipe so I/O thread can effectively send them.

I do not think that it is possible to remove messages from the pipe, at least not in any public API.
Once sendMore() call has succeeded, your message-part is planned for sending, so i would go for "all five will be sent".

I agree with @daveyarwood : the way the message is sent (timeout, blocking or not) may be extremely relevant for us to give you a better focused answer.
